### ISSUE TITLE
Task Cancellation is broken for queued tasks

### ISSUE THREAD

**Issue by ijrsvt** [ray-project/ray#12080]
### What is the problem?
*  After #11051 , tasks that are not yet executing may not be cancellable.
* Tasks cancellation is implemented for tasks that are pending a lease, or are currently running on the worker

## Change needed
* Basically [HandleCancelTask](https://github.com/ray-project/ray/blob/d7c95a4a9065cbad73901b4c1de087837e260316/src/ray/core_worker/core_worker.cc#L2165) should also ask `direct_task_receiver_` to cancel a task (delete it from queue)

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

**Comment by ijrsvt** [comment]
@gabrieleoliaro Hey I saw that you implemented queuing for non-actor tasks (which is **super exciting!**) and I was wondering if you could extend task cancellation to cancel tasks that are queued at the remote backend worker?

**Comment by gabrieleoliaro** [comment]
> @gabrieleoliaro Hey I saw that you implemented queuing for non-actor tasks (which is **super exciting!**) and I was wondering if you could extend task cancellation to cancel tasks that are queued at the remote backend worker?

Thanks for bringing this up, @ijrsvt ! I'll look into this.

**Comment by ijrsvt** [comment]
@gabrieleoliaro Awesome! Please let me know if you have any questions!

**Comment by gabrieleoliaro** [comment]
@ijrsvt I just opened a PR to address this issue! I can't assign people to review my code, so feel free to either assign yourself, or somebody else. Thanks!!

**Comment by ijrsvt** [comment]
@gabrieleoliaro Thanks for doing that so quickly!

**Comment by rkooo567** [comment]
@ijrsvt isn't this done?

**Comment by gabrieleoliaro** [comment]
@rkooo567 just need to fix a couple of build errors

**Comment by starpit** [comment]
Heya. Was this ever fixed? In 1.13.1, I am seeing that if i cancel a job while it is preparing the runtime_env, the job cancellation waits until the runtime_env preparation is done. It seems desirable for job cancellation also to cancel the runtime_env setup?

**Comment by rkooo567** [comment]
@starpit I guess this might be a separate issue. Do you mind creating an issue and tag me there?

**Comment by ijrsvt** [comment]
Yeah the fix for the queued tasks was fixed in the above mentioned PR, then reverted, and then landed in https://github.com/ray-project/ray/pull/13244/

**Comment by cszhu** [comment]
Hello! This P2 issue has seen no activity in at least 2 years. It will be closed in 2 weeks as part of ongoing cleanup efforts.

Please remove the `pending-cleanup` label if you believe this issue should remain open.

Thanks for contributing to Ray!
